File: Replace Words.py
class Trie:
    def __init__(self, is_word=False):
        self.is_word = is_word
        self.paths = {}

    def add(self, i, word):
        if i == len(word):
            self.is_word = True
            return

        c = word[i]
        if c not in self.paths:
            self.paths[c] = Trie()
        self.paths[c].add(i+1, word)

    def search(self, i, word):
        if self.is_word:
            return word[:i]

        if i == len(word) or word[i] not in self.paths:
            return ""

        return self.paths[word[i]].search(i+1, word)

class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        root = Trie()
        for word in dictionary:
            root.add(0, word)

        ans = []
        words = sentence.split(' ')
        for word in words:
            valid_word = root.search(0, word)
            if valid_word:
                ans.append(valid_word)
            else:
                ans.append(word)
        return ' '.join(ans)

File: Replace Words.go
type Trie struct {
    is_word bool
    paths map[byte]*Trie
}

func NewTrie() *Trie {
    return &Trie{
        paths: make(map[byte]*Trie),
    }
}

func (t *Trie) Add(i int, word *string) {
    if i == len(*word) {
        t.is_word = true
        return
    }

    c := (*word)[i]
    if _, ok := t.paths[c]; !ok {
        t.paths[c] = NewTrie()
    }
    t.paths[c].Add(i+1, word)
}

func (t *Trie) Search(i int, word *string) string {
    if t.is_word {
        new_word := (*word)[:i]
        return new_word
    }

    if i == len(*word) {
        return *word
    }

    c := (*word)[i]
    if _, ok := t.paths[c]; !ok {
        return *word
    }

    return t.paths[c].Search(i+1, word)
}

func replaceWords(dictionary []string, sentence string) string {
    root := NewTrie()

    for _, word := range dictionary {
        root.Add(0, &word)
    }

    var ans strings.Builder
    splits := strings.Split(sentence, " ")
    for _, word := range splits {
        valid_word := root.Search(0, &word)
        ans.WriteString(valid_word)
        ans.WriteRune(' ')
    }

    ans_str := ans.String()
    return ans_str[:len(ans_str)-1]
}
